File: apps/web/src/coordinates/transforms.ts
import type {
  CoordinatePatch,
  CoordinateTransform,
  NormalizedStructure,
  Point3D,
  StructureProjection,
} from "../api/types";

type Matrix3 = [Point3D, Point3D, Point3D];

function multiply(left: Matrix3, right: Matrix3): Matrix3 {
  return left.map((row) =>
    right[0].map(
      (_, column) =>
        row[0] * right[0][column] +
        row[1] * right[1][column] +
        row[2] * right[2][column],
    ),
  ) as Matrix3;
}

export function eulerRotationMatrix(rotationDegrees: Point3D): Matrix3 {
  const [x, y, z] = rotationDegrees.map((value) => (value * Math.PI) / 180);
  const rx: Matrix3 = [
    [1, 0, 0],
    [0, Math.cos(x), -Math.sin(x)],
    [0, Math.sin(x), Math.cos(x)],
  ];
  const ry: Matrix3 = [
    [Math.cos(y), 0, Math.sin(y)],
    [0, 1, 0],
    [-Math.sin(y), 0, Math.cos(y)],
  ];
  const rz: Matrix3 = [
    [Math.cos(z), -Math.sin(z), 0],
    [Math.sin(z), Math.cos(z), 0],
    [0, 0, 1],
  ];
  return multiply(rz, multiply(ry, rx));
}

export function coordinateCentroid(
  structure: NormalizedStructure,
  atomIds?: Iterable<number>,
): Point3D {
  const requested = atomIds ? new Set(atomIds) : null;
  const points = structure.atoms.filter((atom) => !requested || requested.has(atom.id));
  if (points.length === 0) throw new Error("A transform pivot requires at least one atom.");
  return points
    .reduce<Point3D>(
      (sum, atom) => [
        sum[0] + atom.coordinates[0] / points.length,
        sum[1] + atom.coordinates[1] / points.length,
        sum[2] + atom.coordinates[2] / points.length,
      ],
      [0, 0, 0],
    );
}

export function previewTransform(
  structure: NormalizedStructure,
  transform: CoordinateTransform,
): CoordinatePatch {
  const selectedIds = transform.selection.atoms
    .filter((atom) => atom.structure_id === transform.entry_id)
    .map((atom) => atom.atom_id);
  const atomIds =
    transform.scope === "structure"
      ? structure.atoms.map((atom) => atom.id)
      : selectedIds;
  const pivot =
    transform.pivot_mode === "custom"
      ? transform.pivot
      : transform.pivot_mode === "selection_centroid"
        ? coordinateCentroid(structure, atomIds)
        : coordinateCentroid(structure);
  if (!pivot) throw new Error("Custom pivot is required.");
  const rotation = eulerRotationMatrix(transform.rotation_degrees);
  const requested = new Set(atomIds);
  return {
    entry_id: transform.entry_id,
    artifact_id: "preview",
    atom_ids: atomIds,
    coordinates: structure.atoms
      .filter((atom) => requested.has(atom.id))
      .map((atom) => {
        const relative: Point3D = [
          atom.coordinates[0] - pivot[0],
          atom.coordinates[1] - pivot[1],
          atom.coordinates[2] - pivot[2],
        ];
        return [
          rotation[0][0] * relative[0] +
            rotation[0][1] * relative[1] +
            rotation[0][2] * relative[2] +
            pivot[0] +
            transform.translation[0],
          rotation[1][0] * relative[0] +
            rotation[1][1] * relative[1] +
            rotation[1][2] * relative[2] +
            pivot[1] +
            transform.translation[1],
          rotation[2][0] * relative[0] +
            rotation[2][1] * relative[1] +
            rotation[2][2] * relative[2] +
            pivot[2] +
            transform.translation[2],
        ];
      }),
  };
}

export function patchStructureProjection(
  projection: StructureProjection,
  patch: CoordinatePatch,
): StructureProjection {
  if (projection.entry_id !== patch.entry_id) return projection;
  const coordinates = new Map(
    patch.atom_ids.map((atomId, index) => [atomId, patch.coordinates[index]]),
  );
  return {
    ...projection,
    structure: {
      ...projection.structure,
      atoms: projection.structure.atoms.map((atom) => {
        const point = coordinates.get(atom.id);
        return point ? { ...atom, coordinates: point } : atom;
      }),
    },
  };
}

File: packages/molweave_core/src/molweave_core/transforms.py
from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

from molweave_core.molecular import Conformer, NormalizedStructureV1

Point3D = tuple[float, float, float]
FloatMatrix = NDArray[np.float64]


class InvalidTransformError(ValueError):
    """Raised when a coordinate transform is undefined or would be a no-op."""


@dataclass(frozen=True, slots=True)
class RigidTransform:
    rotation: tuple[Point3D, Point3D, Point3D]
    translation: Point3D

    @classmethod
    def from_euler(
        cls,
        translation: Sequence[float],
        rotation_degrees: Sequence[float],
        pivot: Sequence[float],
    ) -> RigidTransform:
        translation_array = _vector(translation, "Translation")
        rotation_values = _vector(rotation_degrees, "Rotation")
        pivot_array = _vector(pivot, "Pivot")
        rotation = euler_rotation_matrix(rotation_values.tolist())
        if np.array_equal(translation_array, np.zeros(3)) and np.allclose(
            rotation, np.identity(3), rtol=0, atol=1e-15
        ):
            raise InvalidTransformError("Transform must change at least one coordinate")
        offset = pivot_array - rotation @ pivot_array + translation_array
        return cls(
            rotation=_matrix_tuple(rotation),
            translation=_point(offset),
        )

    def arrays(self) -> tuple[FloatMatrix, NDArray[np.float64]]:
        return (
            np.asarray(self.rotation, dtype=np.float64),
            np.asarray(self.translation, dtype=np.float64),
        )


def euler_rotation_matrix(rotation_degrees: Sequence[float]) -> FloatMatrix:
    """Compose right-handed X, then Y, then Z rotations for column vectors."""

    x_degrees, y_degrees, z_degrees = _vector(rotation_degrees, "Rotation")
    x, y, z = np.deg2rad([x_degrees, y_degrees, z_degrees])
    rotate_x = np.asarray(
        [
            [1.0, 0.0, 0.0],
            [0.0, np.cos(x), -np.sin(x)],
            [0.0, np.sin(x), np.cos(x)],
        ],
        dtype=np.float64,
    )
    rotate_y = np.asarray(
        [
            [np.cos(y), 0.0, np.sin(y)],
            [0.0, 1.0, 0.0],
            [-np.sin(y), 0.0, np.cos(y)],
        ],
        dtype=np.float64,
    )
    rotate_z = np.asarray(
        [
            [np.cos(z), -np.sin(z), 0.0],
            [np.sin(z), np.cos(z), 0.0],
            [0.0, 0.0, 1.0],
        ],
        dtype=np.float64,
    )
    return rotate_z @ rotate_y @ rotate_x


def centroid(
    structure: NormalizedStructureV1,
    atom_ids: Iterable[int] | None = None,
) -> Point3D:
    selected_indices = _atom_indices(structure, atom_ids or (atom.id for atom in structure.atoms))
    coordinates = np.asarray(structure.active_coordinates, dtype=np.float64)
    _validate_coordinates(coordinates)
    return _point(coordinates[selected_indices].mean(axis=0))


def transform_structure(
    structure: NormalizedStructureV1,
    atom_ids: Iterable[int],
    *,
    translation: Sequence[float] = (0.0, 0.0, 0.0),
    rotation_degrees: Sequence[float] = (0.0, 0.0, 0.0),
    pivot: Sequence[float] = (0.0, 0.0, 0.0),
) -> NormalizedStructureV1:
    transform = RigidTransform.from_euler(translation, rotation_degrees, pivot)
    rotation, offset = transform.arrays()
    return apply_rigid_transform(structure, atom_ids, rotation, offset.tolist())


def apply_rigid_transform(
    structure: NormalizedStructureV1,
    atom_ids: Iterable[int],
    rotation: Sequence[Sequence[float]] | FloatMatrix,
    translation: Sequence[float],
) -> NormalizedStructureV1:
    selected_indices = _atom_indices(structure, atom_ids)
    rotation_array = np.asarray(rotation, dtype=np.float64)
    if rotation_array.shape != (3, 3) or not np.all(np.isfinite(rotation_array)):
        raise InvalidTransformError("Rotation matrix must be a finite 3 by 3 matrix")
    translation_array = _vector(translation, "Translation")
    transformed_conformers: list[Conformer] = []
    for conformer in structure.conformers:
        coordinates = np.asarray(conformer.coordinates, dtype=np.float64)
        _validate_coordinates(coordinates)
        changed = coordinates.copy()
        changed[selected_indices] = (
            coordinates[selected_indices] @ rotation_array.T + translation_array
        )
        transformed_conformers.append(
            conformer.model_copy(
                update={
                    "coordinates": [
                        _point(point) for point in changed
                    ]
                }
            )
        )
    active_coordinates = next(
        conformer.coordinates
        for conformer in transformed_conformers
        if conformer.id == structure.active_conformer_id
    )
    atoms = [
        atom.model_copy(update={"coordinates": active_coordinates[index]})
        for index, atom in enumerate(structure.atoms)
    ]
    return structure.model_copy(
        update={"atoms": atoms, "conformers": transformed_conformers}
    )


def _atom_indices(
    structure: NormalizedStructureV1,
    atom_ids: Iterable[int],
) -> NDArray[np.intp]:
    requested = list(atom_ids)
    if not requested:
        raise InvalidTransformError("Transform requires at least one atom")
    if len(set(requested)) != len(requested):
        raise InvalidTransformError("Transform atom IDs must be unique")
    index_by_id = {atom.id: index for index, atom in enumerate(structure.atoms)}
    missing = sorted(set(requested) - index_by_id.keys())
    if missing:
        raise InvalidTransformError(
            f"Transform references missing atom IDs: {', '.join(map(str, missing))}"
        )
    return np.asarray([index_by_id[atom_id] for atom_id in requested], dtype=np.intp)


def _vector(values: Sequence[float], label: str) -> NDArray[np.float64]:
    vector = np.asarray(values, dtype=np.float64)
    if vector.shape != (3,) or not np.all(np.isfinite(vector)):
        raise InvalidTransformError(f"{label} must be a finite 3D vector")
    return vector


def _validate_coordinates(coordinates: FloatMatrix) -> None:
    if (
        coordinates.ndim != 2
        or coordinates.shape[1] != 3
        or not np.all(np.isfinite(coordinates))
    ):
        raise InvalidTransformError("Structure coordinates must be finite 3D points")


def _point(values: Sequence[float] | NDArray[np.float64]) -> Point3D:
    return (float(values[0]), float(values[1]), float(values[2]))


def _matrix_tuple(matrix: FloatMatrix) -> tuple[Point3D, Point3D, Point3D]:
    return (_point(matrix[0]), _point(matrix[1]), _point(matrix[2]))
